File: tools/wiktionary/lib/wikitext.mjs
// tools/wiktionary/lib/wikitext.mjs
//
// Pure text functions. No network, no filesystem, no config - everything here
// is wikitext in, wikitext (or a verdict) out, so the parts that decide what
// we are about to change to somebody else's page are testable on their own.
//
// Everything below was written against the live wikitext of all 114 pages in
// the work queue (536 etymology sections), not against kọ alone. The comments
// record which real page forced which rule, because every one of them looks
// like an over-complication until you meet the page.

// ---------------------------------------------------------------------------
// Markup reduction
// ---------------------------------------------------------------------------

// Templates nest: sun's etymology 4 carries
//   # {{senseid|yo|to cry}} {{lb|yo|with {{l|yo|ẹkún|t=tears}}}} to [[cry]]
// A single non-greedy \{\{.*?\}\} pass eats "{{lb|yo|with {{l|yo|ẹkún|t=tears}}"
// and leaves a stray "}}". Repeatedly removing only the *innermost* templates -
// the ones containing no braces - unwinds the nesting correctly.
export function stripMarkup(text) {
  let out = text;
  let previous = null;
  while (previous !== out) {
    previous = out;
    out = out.replace(/\{\{[^{}]*\}\}/g, '');
  }
  out = out.replace(/\[\[[^\]|]*\|([^\]]*)\]\]/g, '$1'); // [[target|shown]] -> shown
  out = out.replace(/\[\[([^\]]*)\]\]/g, '$1');
  out = out.replace(/'''/g, '').replace(/''/g, '');
  return out.replace(/\s+/g, ' ').trim().replace(/^[;,\s]+|[;,\s]+$/g, '');
}

function contentWords(text) {
  return new Set(
    text
      .toLowerCase()
      .replace(/[^a-z0-9 ]/g, ' ')
      .split(/\s+/)
      .filter(Boolean)
  );
}

// Does a definition line taken from wikitext describe the same thing as one of
// Kaikki's glosses for that section?
//
// Not string equality. Stripping templates removes text that Kaikki *keeps*,
// because Kaikki renders them rather than dropping them:
//
//   wikitext, stripped   "metal iron"
//   kaikki gloss         "metal (in particular) iron"       <- {{q|in particular}}
//   wikitext, stripped   "The tree and its seeds, ..."
//   kaikki gloss         "The tree Picralima nitida and ..." <- {{taxfmt|...}}
//
// Equality rejects 11 of the 502 verifiable sections in the queue, all of them
// correct. Subset-of-content-words in either direction rejects none of them,
// and still fires on 97% of sections when a page is deliberately misaligned by
// one etymology. Returns null when there is nothing left to compare, which is
// the honest answer for the 33 sections whose definition is entirely template.
export function definitionsAgree(fromWikitext, fromKaikki) {
  const a = contentWords(fromWikitext);
  if (!a.size) return null;
  const b = contentWords(fromKaikki);
  if (!b.size) return null;
  const subset = (x, y) => [...x].every((w) => y.has(w));
  return subset(a, b) || subset(b, a);
}

// ---------------------------------------------------------------------------
// Structure
// ---------------------------------------------------------------------------

const ETYMOLOGY_HEADER = /^===\s*Etymology (\d+)\s*===[ \t]*$/gm;

// Split one language section into its numbered etymologies. Returns the text
// before the first etymology header (the L2 header itself, plus anything like
// ===Alternative forms=== that sits above them) so the section can be
// reassembled byte-for-byte.
export function splitEtymologies(sectionWikitext) {
  const headers = [...sectionWikitext.matchAll(ETYMOLOGY_HEADER)];
  if (!headers.length) return { preamble: sectionWikitext, blocks: [] };

  const blocks = headers.map((match, i) => {
    const start = match.index;
    const end = i + 1 < headers.length ? headers[i + 1].index : sectionWikitext.length;
    const raw = sectionWikitext.slice(start, end);
    return {
      number: match[1],
      headerLine: match[0],
      body: raw.slice(match[0].length),
      raw,
      start,
      end,
    };
  });
  return { preamble: sectionWikitext.slice(0, headers[0].index), blocks };
}

export function reassemble(preamble, blocks) {
  return preamble + blocks.map((b) => b.headerLine + b.body).join('');
}

// The definition lines of one etymology block. "# " only: "#:" is a usage
// example, "#*" a quotation, "##" a subsense - none of them is the definition
// we match against.
export function definitionLines(block) {
  return block.body
    .split('\n')
    .filter((line) => /^# [^#*:]/.test(line))
    .map((line) => stripMarkup(line.slice(1)));
}

const ETYMID = /\{\{\s*etymid\s*\|\s*([^|}]*?)\s*\|\s*([^}]*?)\s*\}\}/;

// The name this etymology already carries, or null. Read from live wikitext
// rather than trusted from the build output: the build data is up to a week
// old and anyone can add an anchor in the meantime.
export function existingEtymid(block, langCode = 'yo') {
  const match = block.body.match(ETYMID);
  if (!match || match[1] !== langCode) return null;
  return match[2];
}

// Insert on the line directly after the header. That is where all 19 of the
// etymids currently on queue pages sit, without exception - before the
// etymology prose, not after it.
export function insertEtymid(block, langCode, name) {
  const body = block.body.startsWith('\n') ? block.body.slice(1) : block.body;
  return { ...block, body: `\n{{etymid|${langCode}|${name}}}\n${body}` };
}

// ---------------------------------------------------------------------------
// Names
// ---------------------------------------------------------------------------

// A name becomes the HTML anchor "Yoruba:_<name>" and the value of an idN=
// argument, so it cannot carry the characters that would end the template or
// the argument. Length is not a rule anyone enforces; a name long enough to be
// a sentence is a sign the definition was pasted in whole, so it is worth
// saying rather than blocking.
export function validateName(name) {
  const problems = [];
  if (!name.trim()) problems.push('is empty');
  if (name !== name.trim()) problems.push('has leading or trailing space');
  if (/[|}{=\n\[\]]/.test(name)) problems.push('contains one of | } { = [ ] or a newline');
  if (/^\s*[Ee]tymology\s*\d*\s*$/.test(name)) {
    problems.push('just repeats the section number, which names nothing');
  }
  return problems;
}

File: tools/wiktionary/lib/wikitext.py
"""Reading and changing the wikitext of one language's section.

Pure functions. No network, no Pywikibot site object - text in, text or a
verdict out, so the part that decides what changes on somebody else's page can
be tested on its own.

Parsing is mwparserfromhell and pywikibot.textlib rather than regular
expressions. That is not a preference: sun's etymology 4 carries

    # {{senseid|yo|to cry}} {{lb|yo|with {{l|yo|ekun|t=tears}}}} to [[cry]]

and a non-greedy {{.*?}} eats half of the nested template and leaves a stray
brace. strip_code() unwinds it correctly because it is a parser.
"""

import re

import mwparserfromhell
from pywikibot import textlib

ETYMOLOGY_TITLE = re.compile(r"^=+\s*Etymology\s+(\d+)\s*=+$")


def strip_markup(text):
    """Definition text as a reader sees it, with templates and links resolved."""
    stripped = mwparserfromhell.parse(text).strip_code()
    return re.sub(r"\s+", " ", stripped).strip(" ;,\t")


def definition_lines(content):
    """The definitions of one etymology section.

    "# " only. "#:" is a usage example, "#*" a quotation and "##" a subsense,
    and none of them is the definition we check the numbering against.
    """
    out = []
    for line in content.splitlines():
        if re.match(r"^# [^#*:]", line):
            out.append(strip_markup(line[1:]))
    return out


def language_span(sections, language):
    """Where a language's sections start and end in a flat section list.

    textlib.extract_sections returns every heading on the page in order, so a
    language is its level-2 heading plus everything after it until the next
    level-2. kọ needs this: it carries a Gun section with its own Etymology 1
    and Etymology 2, and touching those would be vandalism.
    """
    start = None
    for i, section in enumerate(sections):
        if section.level == 2:
            if start is not None:
                return start, i
            if section.title.strip("= ") == language:
                start = i
    return (start, len(sections)) if start is not None else (None, None)


def etymology_sections(sections, start, end):
    """One entry per etymology: (heading index, number, index after its last
    subsection).

    The span matters. extract_sections splits at every heading level, so an
    ===Etymology 5=== section's own content is only the etymology prose - the
    definitions are down in ====Verb====, and the pronunciation and derived
    terms are in siblings of that. Everything from the heading until the next
    level-3-or-shallower heading belongs to that etymology.
    """
    found = []
    for i in range(start, end):
        match = ETYMOLOGY_TITLE.match(sections[i].title.strip())
        if match:
            found.append([i, match.group(1), end])
    for position, item in enumerate(found):
        for j in range(item[0] + 1, end):
            if sections[j].level <= 3:
                item[2] = j
                break
    return [tuple(item) for item in found]


def span_content(sections, start, stop):
    """The wikitext of an etymology and everything nested under it."""
    return "".join(sections[i].title + sections[i].content for i in range(start, stop))


def existing_etymid(content, lang_code):
    """The name this etymology already carries, or None.

    Language-scoped on purpose. sun carries {{etymid|fi|conjunction}} and
    {{etymid|fi|pronoun form}} for Finnish alongside the Yoruba one, so a
    template search that ignored the language argument would read a Finnish
    name as ours.
    """
    for template in mwparserfromhell.parse(content).filter_templates():
        if template.name.strip() != "etymid":
            continue
        if len(template.params) >= 2 and str(template.params[0]).strip() == lang_code:
            return str(template.params[1]).strip()
    return None


def existing_senseids(content, lang_code):
    """Every {{senseid}} value in this span.

    etymid and senseid share one namespace: the documentation says IDs "should
    be unique for each invocation of either {{senseid}} or {{etymid}} per
    language section". sun shows the rule being worked around in practice - its
    etymology 2 is named "to roast, burn" because "to roast" and "to burn" were
    already taken by senseids on the same page.
    """
    found = set()
    for template in mwparserfromhell.parse(content).filter_templates():
        if template.name.strip() != "senseid":
            continue
        if len(template.params) >= 2 and str(template.params[0]).strip() == lang_code:
            found.add(str(template.params[1]).strip())
    return found


def insert_etymid(content, lang_code, name):
    """Put the name on the line directly after the heading.

    That is where all 19 of the etymids already on queue pages sit, without
    exception - before the etymology prose, not after it. The section content
    handed to us begins with the newline that followed the heading.
    """
    body = content[1:] if content.startswith("\n") else content
    return "\n{{etymid|%s|%s}}\n%s" % (lang_code, name, body)


def reassemble(parsed, replacements):
    """Rebuild page text with some section contents replaced.

    `replacements` maps a section index to new content. Everything else is
    copied byte for byte, which is what keeps a whole-page save from touching
    any other language on the page.
    """
    out = [parsed.header]
    for i, section in enumerate(parsed.sections):
        out.append(section.title)
        out.append(replacements.get(i, section.content))
    out.append(parsed.footer)
    return "".join(out)


def parse_page(text, site):
    return textlib.extract_sections(text, site)


NAME_FORBIDDEN = re.compile(r"[|}{=\n\[\]]")


def validate_name(name):
    """Why a name cannot be used, if it cannot.

    The name becomes the HTML anchor "Yoruba:_<name>" and the value of an idN=
    argument, so it cannot carry the characters that would end the template or
    the argument.
    """
    problems = []
    # Template:senseid's documentation, which etymid defers to: "Apostrophes
    # have been known not to work for link resolution, at least in some
    # browsers. Thus, please favor (for example) 'sorcerer_s apprentice'".
    # Spaces are fine and explicitly allowed - "This can be anything, and can
    # include spaces" - and the etymid docs' own worked example, bun, uses
    # names like "probably from bottom-referent origins".
    if "'" in name or "\u2019" in name:
        problems.append("contains an apostrophe, which breaks link resolution")
    if not name.strip():
        problems.append("is empty")
    if name != name.strip():
        problems.append("has leading or trailing space")
    if NAME_FORBIDDEN.search(name):
        problems.append("contains one of | } { = [ ] or a newline")
    if re.match(r"^\s*[Ee]tymology\s*\d*\s*$", name):
        problems.append("just repeats the section number, which names nothing")
    return problems
